src/pbaas/MMR.ts

```
@@ -0,0 +1,581 @@
+              let layerIndex = layerNum ? layerNum - 1 : 0;      // layerNum is base 1
+
+              // we are an odd member on the end (even index) and will not hash with the next layer above, we will propagate to its end
+              if ((p < layerSize - 1) || (p & 1)) {
+                if (p & 1) {
+                  // hash with the one before us
+                  if (layerNum > 0) {
+                    let proofHashes = this.peakMerkle[layerIndex][p - 1].hash;
+                    retBranch.branch = retBranch.branch.concat(proofHashes);
+                  }
+                  else {
+                    let proofHashes = this.peaks[p - 1].hash;
+                    retBranch.branch = retBranch.branch.concat(proofHashes);
+                  }
+                }
+                else {
+                  // hash with the one in front of us
+                  if (layerNum > 0) {
+                    let proofHashes = this.peakMerkle[layerIndex][p + 1].hash;
+                    retBranch.branch = retBranch.branch.concat(proofHashes);
+                  }
+                  else {
+                    let proofHashes = this.peaks[p + 1].hash;
+                    retBranch.branch = retBranch.branch.concat(proofHashes);
+                  }
+                }
+              }
+              p >>= 1;
+            }
+
+            // finished
+            break;
+          }
+        }
+      }
+      retBranch.branchType = this.getBranchType();
+      retBranch.nSize = this.size();
+      retBranch.nIndex = pos;
+      retProof.setProof(retBranch);
+      return true;
+    }
+    return false;
+  }
+
+  // return a vector of the bits, either 1 or 0 in each byte, to represent both the size
+  // of the proof by the size of the vector, and the expected bit in each position for the given
+  // position in a Merkle Mountain View of the specified size
+  getProofBits(pos: number, mmvSize: number) {
```

Should throw an error if not implemented imo